x-pack/platform/plugins/shared/streams/server/lib/streams/assets/query/query_client.ts

```diff
@@ -0,0 +1,342 @@
+      config: StreamsConfig;
+    }
+  ) {
+    this.isSignificantEventsEnabled =
+      dependencies.config.experimental?.significantEventsEnabled ?? false;
+  }
+
+  public async syncQueries(stream: string, queries: StreamQuery[]) {
+    if (!this.isSignificantEventsEnabled) {
+      return await this.dependencies.assetClient.syncAssetList(
+        stream,
+        queries.map((query) => ({
+          [ASSET_ID]: query.id,
+          [ASSET_TYPE]: 'query' as const,
+          query,
+        })),
+        'query'
+      );
+    }
+
+    /**
+     * This method is used to synchronize queries/rules for a stream.
+     * It manages the rules associated with these queries, ensuring that any query breaking changes
+     * are handled appropriately:
+     * - If a query is new, it creates a new rule.
+     * - If a query is updated with a breaking change, it removes the old rule and creates a new one.
+     * - If a query is updated without a breaking change, it updates the existing rule.
+     * - If a query is deleted, it removes the associated rule.
+     */
+    const currentQueryLinks = await this.dependencies.assetClient.getAssetLinks(stream, ['query']);
+    const currentIds = new Set(currentQueryLinks.map((link) => link.query.id));
+    const nextIds = new Set(queries.map((query) => query.id));
+
+    const nextQueriesToCreate = queries
+      .filter((query) => !currentIds.has(query.id))
+      .map((query) => toQueryLink(query, stream));
+
+    const currentQueriesToDelete = currentQueryLinks.filter((link) => !nextIds.has(link.query.id));
+
+    const currentQueriesToDeleteBeforeUpdate = currentQueryLinks.filter((link) =>
+      queries.some((query) => query.id === link.query.id && hasBreakingChange(link.query, query))
+    );
+
+    const nextQueriesUpdatedWithBreakingChange = queries
+      .filter((query) =>
+        currentQueryLinks.some(
+          (link) => link.query.id === query.id && hasBreakingChange(link.query, query)
+        )
+      )
+      .map((query) => toQueryLink(query, stream));
+
+    const nextQueriesUpdatedWithoutBreakingChange = queries
+      .filter((query) =>
+        currentQueryLinks.some(
+          (link) => link.query.id === query.id && !hasBreakingChange(link.query, query)
+        )
+      )
+      .map((query) => toQueryLink(query, stream));
+
+    await this.uninstallQueries([...currentQueriesToDelete, ...currentQueriesToDeleteBeforeUpdate]);
+    await this.installQueries(
+      [...nextQueriesToCreate, ...nextQueriesUpdatedWithBreakingChange],
+      nextQueriesUpdatedWithoutBreakingChange,
+      stream
+    );
+
+    await this.dependencies.assetClient.syncAssetList(
+      stream,
+      queries.map((query) => ({
+        [ASSET_ID]: query.id,
+        [ASSET_TYPE]: 'query',
+        query,
+      })),
+      'query'
+    );
+  }
+
+  public async upsert(stream: string, query: StreamQuery) {
+    const { assetClient } = this.dependencies;
+
+    if (!this.isSignificantEventsEnabled) {
```

Comment:
Shouldn't we throw here? If we don't, what happens to the queries when streams is enabled after the fact?

Reply:
Actually, the do nothing solution would break all current tests... 
I think if we continue storing the query asset links it will be fine. Worst case a customer has stored some queries, and then enable sig events, then they would have to simply upsert their stream again to get the rules created.